Introduce the BlockClock control

shaavan opened the pull request:
- This PR introduces the Block clock control and adds it to the stub.qml to display.
- This PR implements the syncing state of the Block Clock.
- The implementation follows the [design guide](https://bitcoincore.app/block-clock/) and uses [this](https://stupefied-jones-dd209f.netlify.app/) prototype as the reference.

**Screenshot:**

![Screenshot from 2022-08-09 17-43-03](https://user-images.githubusercontent.com/85434418/183645698-460cce48-9fb2-43de-89e5-1309458a7833.png)


Things to note for the current implementation:

Since this is the prototype for the Block clock, I focused on the implementation part and took some liberties with the nuances, which I would like to document here.

1. The color values for the: **Remaining time** text and **Background color of the block clock ring** didn't match the exact color values in the Theme.qml file. So I have used the closet matching Theme.color values to avoid hardcoded values and maintain consistency between light and dark mode.
2. Since there was no "yellow" color in the Theme.qml, I have used a hardcoded value for the progress indicator.
3. I have used the "bitcoin app" icon instead of the Black and White bitcoin icon, as this required additions to the resources and would be best done in a separate PR.
4. I have used static dots as the peer indicator. PR #127 introduces an implementation of the peer indicator and has a largely positive review. This PR could be updated with #127's implementation when that gets merged.

jarolrod commented on src/Makefile.qt.include:
```diff
@@ -312,6 +312,7 @@ QML_RES_QML = \
   qml/components/StorageLocations.qml \
   qml/components/StorageOptions.qml \
   qml/components/StorageSettings.qml \
+  qml/controls/BlockClock.qml \
```
this is a component. not a control

shaavan replied:
The next update of my PR intends to make the `BlockClock.qml` file a purely abstract structure and introduce a component file that uses it to paint different states of the block clock as needed. So I think moving `BlockClock.qml` to components now would be a wasted effort.

Should you still think I shall move the `BlockClock.qml` to controls, or if there is a better approach for achieving the final BlockClock state, I would love to hear that.

jarolrod commented on src/qml/controls/BlockClock.qml:
```diff
@@ -0,0 +1,106 @@
+        }
+    }
+
+    Canvas {
```
instead of canvas, this should be the simpler ShapePath: https://doc.qt.io/qt-6/qml-qtquick-shapes-shapepath.html

shaavan replied:
I looked through using the ShapePath to paint the BlockClock. However, using this would require us to introduce a new dependency, `qml-module-qtquick-shapes` for building the GUI.

So I think this suggestion needs further discussion before implementing it.

A friendly ping, @hebasto

What do you think about this suggestion?

promag replied:
Unfortunately, from my experiments, `ShapePath` and related items have issues. The most annoying issue happens with small curves where tessellation yields bad curve smoothness. Another issue is that with complex drawings you end up with lots of items and bindings everywhere, which is also not that performant.

Going forward a stack/composition of Canvas is the best option, like photoshop layers. For instance, the Canvas on the bottom could be static or updated less often than the one on the top.

jarolrod commented on src/qml/nodemodel.cpp:
```diff
@@ -44,6 +120,13 @@ void NodeModel::initializeResult([[maybe_unused]] bool success, interfaces::Bloc
+
+    const CBlockIndex* tip;
+    node::NodeContext* context = m_node.context();
+    ChainstateManager* chainman = context->chainman.get();
```
this is too low level of an object to be handling here when we have `interfaces::chain`

shaavan replied:
Let me take a look into replacing this low-level object handling with interfaces::chain

jarolrod commented on src/qml/nodemodel.h:
```diff
@@ -25,15 +29,24 @@ class NodeModel : public QObject
     double verificationProgress() const { return m_verification_progress; }
     void setVerificationProgress(double new_progress);
+    QVariantList blockTimeList() const { return m_block_time_list; }
+    bool setBlockTimeList(int new_block_time);
```
why should the nodemodel, a proxy class for our internal node, keep something like a blocklist?

shaavan replied:
Considering your previous suggestion about using the `interfaces::chain`, I think I can create a separate .cpp file handling chain-related parameters such as blockTimeList

jarolrod commented on src/qml/controls/BlockClock.qml:
```diff
@@ -0,0 +1,160 @@
+import QtQuick.Controls 2.15
+import QtQuick.Layouts 1.15
+
+Item {
```
This is the implementation of the block clock, and the file residing at `src/qml/components/BlockClockComponent.qml` is just a controller for this. A better architecture is to have a controller that loads the different states the block clock can be in.

You'd be able to split up this file, as it is very long, between a simple circular progress bar that shows the verification progress while in IBD, then another dial like progress bar for the blocks. Then load these into the controller.

shaavan replied:
Let me try rephrasing your suggestion to ensure I understood it correctly.
You suggest I create a separate `component` file for each block clock state, create a controller file in `qml/control`, and switch to loading different block clock component files based on the requirement.

Have I understood it correctly?

jarolrod commented on src/qml/controls/BlockClock.qml:
```diff
@@ -0,0 +1,160 @@
+            sourceSize.height: 30
+            Layout.bottomMargin: 15
+        }
+        Text {
```
isn't this recreating our header component?

jarolrod commented on src/qml/nodemodel.cpp:
```diff
@@ -28,12 +29,87 @@ void NodeModel::setBlockTipHeight(int new_height)
+    return true;
+}
+
+void NodeModel::setBlockTimeListInitial(const CBlockIndex* pblockindex)
```
Yeah this is definitely too low level for us to be passing around here

shaavan replied:
Let me experiment with `interfaces::chain` and devise an approach that avoids low-level object handling.

jarolrod reviewed (commented):
Branch doesn't build

shaavan commented:
Updated from 621eae1074365116d5bfd5e58eff2696eaf6a846 to 1f134fdb3bcc29213b55658a69d1d8bf72079554 ([pr148.02](https://github.com/shaavan/gui-qml/commits/pr148.02) -> [pr148.03](https://github.com/shaavan/gui-qml/commits/pr148.03), [diff](https://github.com/shaavan/gui-qml/compare/pr148.02..pr148.03))
Addressed @jarolrod comment

Update:
- Made changes to fix the problem of the branch not building. Thanks, @jarolrod, for bringing that to notice and helping figure out the root of the problem.

shaavan commented:
Updated from 23f8a833b57bbdad3177ff23751ec0cdfb8d5b09 to 621eae1074365116d5bfd5e58eff2696eaf6a846 ([pr148.03](https://github.com/shaavan/gui-qml/commits/pr148.03) -> [pr148.04](https://github.com/shaavan/gui-qml/commits/pr148.04))

Changes:
- Introduced a new chainmodel class to use the interfaces::chain and directly use the chain info without going too low level in the code.
- The new Block Clock displays the timezone-dependent block time data.
- Added property to Header.qml so it can be used directly in the BlockClock.qml file.

Here is the demo of the latest iteration of the block clock:

https://user-images.githubusercontent.com/85434418/204341167-fe8cb5a4-7189-4cbd-b2f8-1db93be3cf3d.mov

shaavan commented:
Updated from 621eae1074365116d5bfd5e58eff2696eaf6a846 to 2a5e9983bebab10ce7be79eebdaf6104a992ca32 ([pr148.04](https://github.com/shaavan/gui-qml/commits/pr148.04) -> [pr148.05](https://github.com/shaavan/gui-qml/commits/pr148.05))

Changes: 
1. Rebased over main.
2. Added the **Connecting** and **Paused** state for the block clock.
3. Integrated the first three commits of #127. This allows the PeerIndicator dots to be animated and correctly show the peer connection state.


**Demo:**

https://user-images.githubusercontent.com/85434418/207072355-c7371de4-fdd5-4dbe-bfee-db74e9a61818.mov

GBKS commented:
Thanks for the updates. It looks like the artifacts are not building due a lint issue (the Android build is only 22 bytes). Could you please address this? I'd love to get hands-on with this update.

hebasto commented on .vscode/settings.json:
```diff
@@ -0,0 +1,7 @@
+{
+    "files.associations": {
+        "__split_buffer": "cpp",
+        "deque": "cpp",
+        "list": "cpp"
+    }
+}
```
Why is this file required?

shaavan replied:
The file has been removed in 0f5461e

hebasto reviewed (commented):
To fix the lint task, suggesting the patch as follows:
```diff
diff --git a/src/qml/chainmodel.h b/src/qml/chainmodel.h
index 9ab25f35c..2943c9c16 100644
--- a/src/qml/chainmodel.h
+++ b/src/qml/chainmodel.h
@@ -2,8 +2,8 @@
 // Distributed under the MIT software license, see the accompanying
 // file COPYING or http://www.opensource.org/licenses/mit-license.php.
 
-#ifndef BITCOIN_QML_CHAIN_MODEL_H
-#define BITCOIN_QML_CHAIN_MODEL_H
+#ifndef BITCOIN_QML_CHAINMODEL_H
+#define BITCOIN_QML_CHAINMODEL_H
 
 #include <interfaces/chain.h>
 
@@ -47,4 +47,4 @@ private:
     interfaces::Chain& m_chain;
 };
 
-#endif // BITCOIN_QML_CHAIN_MODEL_H
\ No newline at end of file
+#endif // BITCOIN_QML_CHAINMODEL_H
```

jarolrod commented on src/qml/controls/BlockClock.qml:
```diff
@@ -0,0 +1,157 @@
+        anchors.centerIn: root
+        Image {
+            Layout.alignment: Qt.AlignCenter
+            source: "image://images/app"
```
This should be using the bitcoin circle icon introduced in https://github.com/bitcoin-core/gui-qml/pull/176. That's what is specified in the [design files](https://www.figma.com/file/GaCoOSNHB2yMB9ThiDtred/Bitcoin-Core-App?node-id=3259%3A73676&t=pjV9bY1q9KckaP4Q-0)

shaavan replied:
Addressed in 0f5461e.

shaavan commented:
Updated from 2a5e9983bebab10ce7be79eebdaf6104a992ca32 to 0f5461eef14e3ac0739d9b3e3448dd78c94ebff7 ([pr148.05](https://github.com/shaavan/gui-qml/commits/pr148.05) -> [pr148.06](https://github.com/shaavan/gui-qml/commits/pr148.06))

Addressed @hebasto @jarolrod @GBKS comments.

Changes:

1. Switch the Bitcoin app icon for the Bitcoin-circle icon.
2. Fixed `QObject::moveToThread: Cannot move objects with a parent` warning.
3. Correction Font size and margins according to [design docs](https://www.figma.com/file/ek8w3n3upbluw5UL2lGhRx/Bitcoin-Core-App-Design?node-id=616%3A0&t=poh7bpeInsyETLvD-0).
4. Removed .vscode/settings.json file.
5. Applied the patch suggested by @hebasto to fix the lint warning.

**Screenshots:**

|pr148.05|pr148.06|
|---------|---------|
|<img width="340" alt="Screenshot 2022-12-15 at 11 51 04 PM" src="https://user-images.githubusercontent.com/85434418/207944317-16fa188c-2874-473f-8388-c368771eaf87.png">|<img width="340" alt="Screenshot 2022-12-15 at 11 48 44 PM" src="https://user-images.githubusercontent.com/85434418/207944388-604e628d-2c75-4651-9d7d-c6a57e099947.png">|
|<img width="340" alt="Screenshot 2022-12-15 at 11 52 02 PM" src="https://user-images.githubusercontent.com/85434418/207944464-ad1469a8-e3d1-4eca-ac27-dc601e9e0b23.png">|<img width="340" alt="Screenshot 2022-12-15 at 11 49 19 PM" src="https://user-images.githubusercontent.com/85434418/207944492-aedc46b3-d08c-4792-9a18-d159675c380c.png">|
|<img width="340" alt="Screenshot 2022-12-15 at 11 51 32 PM" src="https://user-images.githubusercontent.com/85434418/207944572-848af8c6-cc8b-4fec-ba30-daf39941bbd3.png">|<img width="340" alt="Screenshot 2022-12-15 at 11 49 51 PM" src="https://user-images.githubusercontent.com/85434418/207944608-7cb2eaff-cdd6-441b-a2f6-8e3b1c1a7488.png">|
|<img width="340" alt="Screenshot 2022-12-15 at 11 51 30 PM" src="https://user-images.githubusercontent.com/85434418/207944670-f2dda596-aa9e-4a2f-a79a-86b6a4407c2d.png">|<img width="340" alt="Screenshot 2022-12-15 at 11 49 43 PM" src="https://user-images.githubusercontent.com/85434418/207944700-c3cb3bd5-2a0e-4d08-9980-56e6b9dd983c.png">|

shaavan commented on src/qml/chainmodel.cpp:
```diff
@@ -0,0 +1,75 @@
+    timer->start(1000);
+
+    QThread* timer_thread = new QThread;
+    timer->moveToThread(timer_thread);
```
Fixed in 0f5461e

shaavan commented:
Updated from 0f5461eef14e3ac0739d9b3e3448dd78c94ebff7 to 0fbe6c809eadabde0ca4c6f443dea60feae3fb34 ([pr148.06](https://github.com/shaavan/gui-qml/commits/pr148.06) -> [pr148.07](https://github.com/shaavan/gui-qml/commits/pr148.07))

- Rebased the branch over the main.

shaavan commented:
Updated from [0fbe6c8 ](https://github.com/bitcoin-core/gui-qml/commit/0fbe6c809eadabde0ca4c6f443dea60feae3fb34)to [29f019e](https://github.com/bitcoin-core/gui-qml/commit/29f019e4a5278a4e7b5aaf56a49bee309bcc4d2a)

- Fixed the trailing whitespace lint failure.

shaavan commented:
Updated from 29f019e4a5278a4e7b5aaf56a49bee309bcc4d2a to 8173bf6838997dce6dc6e908a54801fdcb5e0ad8 ([pr148.08](https://github.com/shaavan/gui-qml/commits/pr148.08) -> [pr148.09](https://github.com/shaavan/gui-qml/commits/pr148.09))

- Moved the simplified block_clock changes from #212 to this PR to preserve the comment history for the block_clock and prevent splitting of discussion in two separate PRs.

Differences as explained in #212 Description:

> 1. The PeersIndicator has been replaced with plain rectangles to keep the UI intact.
> 2. The conns property representing if the number of connections is greater than 0, has been set to always positive. With this > change, the "connecting" page, though implemented, becomes inaccessible through the UI.

johnny9 commented on src/qml/components/BlockClockComponent.qml:
```diff
@@ -0,0 +1,64 @@
+                ringProgress: blockList[0]
+                header: Math.floor(nodeModel.blockTipHeight / 1000) + "," + nodeModel.blockTipHeight % 1000
+                subText: "Blocktime"
+                blockList: chainModel.timeRatioList
```
QML error appears when shutting down.

```
2022-12-17T03:52:37Z Shutdown: In progress...
2022-12-17T03:52:37Z torcontrol thread exit
2022-12-17T03:52:37Z addcon thread exit
2022-12-17T03:52:37Z msghand thread exit
2022-12-17T03:52:37Z net thread exit
2022-12-17T03:52:39Z opencon thread exit
2022-12-17T03:52:39Z DumpAnchors: Flush 0 outbound block-relay-only peer addresses to anchors.dat started
2022-12-17T03:52:39Z DumpAnchors: Flush 0 outbound block-relay-only peer addresses to anchors.dat completed (0.01s)
2022-12-17T03:52:39Z scheduler thread exit
2022-12-17T03:52:39Z Writing 0 unbroadcast transactions to disk.
2022-12-17T03:52:39Z Dumped mempool: 3.7e-06s to copy, 0.00448635s to dump
2022-12-17T03:52:39Z Shutdown: done
2022-12-17T03:52:39Z GUI: qrc:/qml/components/BlockClockComponent.qml:34: TypeError: Cannot read property 'timeRatioList' of null
```

shaavan replied:
Addressed in bcbc80b6666e43aa07ee46e09289b4a34f3fa30b

johnny9 commented on src/qml/controls/BlockClock.qml:
```diff
@@ -0,0 +1,175 @@
+            else {
+                var del = 0.0025
+                // Paint Block time points
+                for (var i=1; i<parent.blockList.length - 1; i++) {
```
The following error appears when going into the paused state.

```
2022-12-17T03:50:17Z GUI: qrc:/qml/controls/BlockClock.qml:96: TypeError: Value is undefined and could not be converted to an object
```

johnny9 replied:
Small style comment, operators should have space before/after for easier readability.

shaavan replied:
Addressed in 20b4ed55f1b9417fa0a1457736be1b6c1c54f1dd

johnny9 commented on src/qml/pages/node/NodeRunner.qml:
```diff
@@ -18,26 +18,9 @@ Page {
+
+        Component.onCompleted: nodeModel.startNodeInitializionThread();
+
+        BlockClockComponent{}
```
```
GUI: qrc:/qml/pages/node/NodeRunner.qml:24:9: QML BlockClockComponent: Detected anchors on an item that is managed by a layout. This is undefined behavior; use Layout.alignment instead.
```
Should be able to remove the ColumnLayout containing the BlockClockComponent and just define the BlockClockComponent as the content to this Page.  When you do this, also remove the `anchors.fill: parent` and `spacing: 0` in this file.

shaavan replied:
Addressed in bcbc80b6666e43aa07ee46e09289b4a34f3fa30b.

johnny9 commented on src/qml/chainmodel.h:
```diff
@@ -0,0 +1,50 @@
+    interfaces::Chain& m_chain;
+};
+
+#endif // BITCOIN_QML_CHAINMODEL_H
```
"No newline at end of file"

shaavan replied:
Addressed in bcbc80b6666e43aa07ee46e09289b4a34f3fa30b

johnny9 commented on src/qml/chainmodel.cpp:
```diff
@@ -0,0 +1,75 @@
+        m_time_ratio_list.erase(m_time_ratio_list.begin() + 1, m_time_ratio_list.end());
+    }
+    m_time_ratio_list[0] = currentTimeRatio;
+}
```
"No newline at end of file"

shaavan replied:
Addressed in bcbc80b6666e43aa07ee46e09289b4a34f3fa30b

johnny9 reviewed (changes requested):
Views look nice and render correctly.  Tested pause/unpause as well as a full sync on signet at got the following results.

It looks great.  Tested pause/resume as well as a sync on signet and here's what I got.

|![Screenshot_2022-12-16_22-18-31](https://user-images.githubusercontent.com/985648/208221768-bbf55e20-d529-4472-94cd-a7115f71270e.png)|![Screenshot_2022-12-16_22-19-02](https://user-images.githubusercontent.com/985648/208221777-46e59c67-fe80-4d5d-9877-8a247f069b78.png)|![Screenshot_2022-12-16_22-25-01](https://user-images.githubusercontent.com/985648/208221788-9deabdf1-389a-4a61-b264-fbabf8a28d2c.png)|![Screenshot_2022-12-16_22-26-42](https://user-images.githubusercontent.com/985648/208221794-9d770b93-b6e1-4f81-b835-a4223844d6f0.png)|![Screenshot_2022-12-16_22-28-51](https://user-images.githubusercontent.com/985648/208221797-4ebe4269-c89a-4301-9f15-ab5dc71f1f93.png)|
|---|---|---|---|---|

A few QML errors when running that need to be resolved before merging. I've left comments for those errors at the corresponding line numbers.

`tail -f debug.log | grep GUI` to watch for issues when running bitcoin-qt.

shaavan commented:
Updated from 8173bf6838997dce6dc6e908a54801fdcb5e0ad8 to bcbc80b6666e43aa07ee46e09289b4a34f3fa30b ([pr148.09](https://github.com/shaavan/gui-qml/commits/pr148.09) -> [pr148.10](https://github.com/shaavan/gui-qml/commits/pr148.10))

Addressed @johnny9 comments.

Changes:

1. Fixed Type Errors as pointed out in [this](https://github.com/bitcoin-core/gui-qml/pull/148#pullrequestreview-1222038049) comment.
2. Added space before and after operator in BlockClock.qml
3. Added a new line at the end of the file in chainmodel.cpp and chainmodel.h.
4. Removed the redundant ColumnLayout in Noderunner.qml.

hebasto commented:
The CI lint task error:
```
fatal: Invalid revision range fd4e57f63654e2d8e7d0de0aafc335715a53a5cc..c4e1ce262ffb5e6956a7b912f2dcf7560823682e
```

looks a bit weird as I failed to find a branch the c4e1ce262ffb5e6956a7b912f2dcf7560823682e commit belongs to.

I hope rebasing will help.

@shaavan And please make sure your local branch does not contain any merge commits.

shaavan commented:
Updated from bcbc80b6666e43aa07ee46e09289b4a34f3fa30b to 68fa33b202e528ff0eae839f7163be8a7f939392  ([pr148.10](https://github.com/shaavan/gui-qml/commits/pr148.10) -> [pr148.11](https://github.com/shaavan/gui-qml/commits/pr148.11))

- Rebased PR over current main.

johnny9 commented on src/qml/controls/BlockClock.qml:
```diff
@@ -0,0 +1,177 @@
+
+            else {
+                var del = 0.0025
+                if (parent.blockList.length) {
```
Still getting `2022-12-19T23:10:24Z GUI: qrc:/qml/controls/BlockClock.qml:95: TypeError: Value is undefined and could not be converted to an object` when going to the paused state. i think doing `if (parent.blockList) {` should resolve it.

shaavan replied:
Addressed in 20b4ed55f1b9417fa0a1457736be1b6c1c54f1dd

Thanks for all of your reviews!

johnny9 reviewed (changes requested):
QML error when going into the paused state still exists. Other issues are resolved.

shaavan commented:
Updated from 68fa33b202e528ff0eae839f7163be8a7f939392 to 20b4ed55f1b9417fa0a1457736be1b6c1c54f1dd ([pr148.11](https://github.com/shaavan/gui-qml/commits/pr148.11) -> [pr148.12](https://github.com/shaavan/gui-qml/commits/pr148.12), [diff](https://github.com/shaavan/gui-qml/compare/pr148.11..pr148.12))

Addressed @johnny9 comment.

- Fixed `TypeError: Value is undefined and could not be converted to an object` error.

- Also addressed one formatting issue, present in displaying Block Counter.

**Explaination:**

For example, for the block count of **122010**, difference:

|pr148.11|pr148.12|
|--------|---------|
|<img width="340" alt="Screenshot 2022-12-20 at 5 54 00 PM" src="https://user-images.githubusercontent.com/85434418/208671008-09b955f0-fbf2-434d-8cbe-11f12acbb3cc.png">|<img width="340" alt="Screenshot 2022-12-20 at 6 16 46 PM" src="https://user-images.githubusercontent.com/85434418/208671046-71fe0d9c-0b9d-4856-9ec0-b9059059fb76.png">|

shaavan commented:
Updated from 20b4ed55f1b9417fa0a1457736be1b6c1c54f1dd to abb0bef2460b4ba98cf02758dedd052efe4ec722 ([pr148.12](https://github.com/shaavan/gui-qml/commits/pr148.12) -> [pr148.13](https://github.com/shaavan/gui-qml/commits/pr148.13), [diff](https://github.com/shaavan/gui-qml/compare/pr148.12..pr148.13))

- Modified `timeRatioList` so that blockClock start displaying from the "0" point.

**Screenshot:**

|pr148.12|pr148.13|
|--------|---------|
|<img width="340" alt="Screenshot 2022-12-20 at 6 28 41 PM" src="https://user-images.githubusercontent.com/85434418/208673687-faf8bb69-e717-4cf0-bb8a-b545a5fbb721.png">|<img width="340" alt="Screenshot 2022-12-20 at 6 27 57 PM" src="https://user-images.githubusercontent.com/85434418/208673800-de9da7c2-2836-478d-88db-38ba85e640e9.png">|

johnny9 reviewed (commented):
Good work on this Shashwat. All of the QML errors appear to be resolved now. Views render as expected

hebasto commented on src/node/interfaces.cpp:
```diff
@@ -514,6 +514,11 @@ class ChainImpl : public Chain
+    int64_t getBlockTime(int height) override
+    {
+        LOCK(::cs_main);
+        return Assert(chainman().ActiveChain()[height])->GetBlockTime();
```
fec6a0e9c7c4d98a88628ab57ce33144923da85d

In `Chain` interface, probably better to use a type-safe `CBlockIndex::Time()` https://github.com/bitcoin-core/gui-qml/blob/fd4e57f63654e2d8e7d0de0aafc335715a53a5cc/src/chain.h#L279-L282

Anyway, it will take a broader discussion when it will about to merge into the main repo.

shaavan replied:
Sure, let me look through it.

> it will take a broader discussion when it will about to merge into the main repo.

Let me include that, along with other future suggestions for this file.

hebasto commented on src/qml/chainmodel.h:
```diff
@@ -0,0 +1,50 @@
+private:
+    QVariantList m_time_ratio_list{0.0};
+
+    QTimer* timer;
```
a287f9e44660b80349defa8f34f14c8b9f0d8b78
Naming convention:
```suggestion
    QTimer* m_timer;
```

shaavan replied:
Addressed in 67d12a67536b636bbda3f5bb8d1e62cc5fa66d65

Since this timer didn't need to be a global variable, I changed it to a local variable.

shaavan commented on src/qml/chainmodel.cpp:
```diff
@@ -0,0 +1,76 @@
+ChainModel::ChainModel(interfaces::Chain& chain)
+    : m_chain{chain}
+{
+    timer = new QTimer();
+    connect(timer, &QTimer::timeout, this, &ChainModel::setCurrentTimeRatio);
+    timer->start(1000);
+
+    QThread* timer_thread = new QThread;
+    timer->moveToThread(timer_thread);
+    timer_thread->start();
```
The **currentTimeRatio** represents the ratio of 12 hours that has been passed.
So, for example, at 4 am or pm, the currenTimeRatio would be 0.3.

We need to update this information in the blockClock each second so that the blockClock always shows the correct blockTime.

So that's why I created a separate 1-sec timer thread to call this function each second and update the **currentTimeRatio**.

shaavan commented:
Updated from abb0bef2460b4ba98cf02758dedd052efe4ec722 to 67d12a67536b636bbda3f5bb8d1e62cc5fa66d65 ([pr148.13](https://github.com/shaavan/gui-qml/commits/pr148.13) -> [pr148.14](https://github.com/shaavan/gui-qml/commits/pr148.14), [diff](https://github.com/shaavan/gui-qml/compare/pr148.13..pr148.14))

Addressed @hebasto comments.

Changes:

- Changed 1-sec QTimer from a global variable to a local variable.
- Changed names of variables mentioned in [this](https://github.com/bitcoin-core/gui-qml/pull/148#discussion_r1053596002) comment from camel-style naming to snake-style naming. For full diff, check [diff](https://github.com/shaavan/gui-qml/compare/pr148.13..pr148.14).
- Added descriptive comment for `m_time_ratio_list`, and the values it stores.

shaavan commented on src/qml/chainmodel.cpp:
```diff
@@ -0,0 +1,76 @@
+int ChainModel::timestampAtMeridian()
+{
+    int secsSinceMeridian = (QTime::currentTime().msecsSinceStartOfDay() / 1000) % SECS_IN_12_HOURS;
+    int currentTimestamp = QDateTime::currentSecsSinceEpoch();
```
Addressed in 67d12a67536b636bbda3f5bb8d1e62cc5fa66d65

shaavan commented on src/qml/chainmodel.cpp:
```diff
@@ -0,0 +1,76 @@
+
+void ChainModel::setTimeRatioListInitial()
+{
+    int timeAtMeridian = timestampAtMeridian();
```
Addressed in 67d12a67536b636bbda3f5bb8d1e62cc5fa66d65

shaavan commented on src/qml/chainmodel.cpp:
```diff
@@ -0,0 +1,76 @@
+void ChainModel::setCurrentTimeRatio()
+{
+    int secsSinceMeridian = (QTime::currentTime().msecsSinceStartOfDay() / 1000) % SECS_IN_12_HOURS;
+    double currentTimeRatio = double(secsSinceMeridian) / SECS_IN_12_HOURS;
```
Addressed in 67d12a67536b636bbda3f5bb8d1e62cc5fa66d65

shaavan commented on src/qml/chainmodel.h:
```diff
@@ -0,0 +1,50 @@
+    void timeRatioListChanged();
+
+private:
+    QVariantList m_time_ratio_list{0.0};
```
Addressed in 67d12a67536b636bbda3f5bb8d1e62cc5fa66d65

hebasto commented:
Closing in favor of #220.
